routers/web/repo/actions/badge.go

```diff
@@ -0,0 +1,177 @@
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/util"
+
+	"golang.org/x/image/font"
+	"golang.org/x/image/font/basicfont"
+)
+
+type Badge struct {
+	Label         string
+	Message       string
+	Color         string
+	Width         int
+	LabelWidth    int
+	MessageWidth  int
+	LabelX        int
+	MessageX      int
+	LabelLength   int
+	MessageLength int
+	MessageColor  string
+	MessageShadow string
+	FontSize      int
+}
+
+const (
+	defaultOffset   = 9
+	defaultSpacing  = 0
+	defaultFontSize = 11
+)
+
+var drawer = &font.Drawer{
+	Face: basicfont.Face7x13,
+}
+
+var statusColorMap = map[actions_model.Status]string{
+	actions_model.StatusSuccess:   "#4c1",    // Green
+	actions_model.StatusSkipped:   "#dfb317", // Yellow
+	actions_model.StatusUnknown:   "#97ca00", // Light Green
+	actions_model.StatusFailure:   "#e05d44", // Red
+	actions_model.StatusCancelled: "#fe7d37", // Orange
+	actions_model.StatusWaiting:   "#dfb317", // Yellow
+	actions_model.StatusRunning:   "#dfb317", // Yellow
+	actions_model.StatusBlocked:   "#dfb317", // Yellow
+}
+
+func GetDefaultBranchWorkflowBadge(ctx *context.Context) {
+	workflowFile := ctx.Params("workflow_name")
+	branchName := git.RefNameFromBranch(ctx.Repo.Repository.DefaultBranch).String()
+	badge, err := getWorkflowBadge(ctx, workflowFile, branchName)
+	if err != nil {
+		if errors.Is(err, util.ErrNotExist) {
+			ctx.NotFound("Not found", fmt.Errorf("%s not found", workflowFile))
+			return
+		}
+		ctx.ServerError("GetWorkflowBadge", err)
+		return
+	}
+	ctx.Data["Badge"] = badge
+	ctx.HTML(http.StatusOK, "shared/actions/runner_badge")
+}
+
+func GetWorkflowBadge(ctx *context.Context) {
+	workflowFile := ctx.Params("workflow_name")
+	branchName := ctx.Params("branch_name")
+	badge, err := getWorkflowBadge(ctx, workflowFile, branchName)
+	if err != nil {
+		if errors.Is(err, util.ErrNotExist) {
+			ctx.NotFound("Not found", fmt.Errorf("%s not found", workflowFile))
+			return
+		}
+		ctx.ServerError("GetWorkflowBadge", err)
+		return
+	}
+	ctx.Data["Badge"] = badge
+	ctx.HTML(http.StatusOK, "shared/actions/runner_badge")
+}
+
+func getWorkflowBadge(ctx *context.Context, workflowFile, branchName string) (Badge, error) {
+	run, err := actions_model.GetRepoBranchLastRun(ctx, ctx.Repo.Repository.ID, branchName, workflowFile)
+	if err != nil {
+		return Badge{}, err
+	}
+
+	color, ok := statusColorMap[run.Status]
+	if !ok {
+		return Badge{}, fmt.Errorf("unknown status %d", run.Status)
+	}
+
+	extension := filepath.Ext(workflowFile)
+	workflowName := strings.TrimSuffix(workflowFile, extension)
+	return generateBadge(workflowName, run.Status.String(), color), nil
+}
+
+// utils for badge generation -------------------------------------
+
+// generateBadge generates badge with given template
+func generateBadge(label, message, color string) Badge {
+	gF := float64(defaultOffset)
+	lW := float64(drawer.MeasureString(label)>>6) + gF
+	mW := float64(drawer.MeasureString(message)>>6) + gF
```

Comment:
Is it right to do so (guessing font size)?

The SVG uses `font-family="Geneva,DejaVu Sans,sans-serif"`, do you really know its width on client side?

Reply:
I think there is no way knowing how it would render on a client, but as long as these approximations work on common clients, I can accept them.